Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 11264)
+++ libavcodec/Makefile	(working copy)
@@ -140,6 +140,7 @@
 OBJS-$(CONFIG_PNG_ENCODER)             += png.o pngenc.o
 OBJS-$(CONFIG_PPM_ENCODER)             += pnmenc.o pnm.o
 OBJS-$(CONFIG_PTX_DECODER)             += ptx.o
+OBJS-$(CONFIG_QCELP_DECODER)           += qcelpdec.o
 OBJS-$(CONFIG_QDM2_DECODER)            += qdm2.o mdct.o fft.o mpegaudiodec.o mpegaudiodecheader.o mpegaudio.o mpegaudiodata.o
 OBJS-$(CONFIG_QDRAW_DECODER)           += qdrw.o
 OBJS-$(CONFIG_QPEG_DECODER)            += qpeg.o
@@ -325,6 +326,7 @@
 OBJS-$(CONFIG_MPEGAUDIO_PARSER)        += mpegaudio_parser.o mpegaudiodecheader.o mpegaudiodata.o
 OBJS-$(CONFIG_MPEGVIDEO_PARSER)        += mpegvideo_parser.o
 OBJS-$(CONFIG_PNM_PARSER)              += pnm_parser.o pnm.o
+OBJS-$(CONFIG_QCELP_PARSER)            += qcelp_parser.o
 OBJS-$(CONFIG_VC1_PARSER)              += vc1_parser.o
 
 OBJS-$(CONFIG_DUMP_EXTRADATA_BSF)      += dump_extradata_bsf.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 11264)
+++ libavcodec/allcodecs.c	(working copy)
@@ -189,6 +189,7 @@
     REGISTER_DECODER (MPC7, mpc7);
     REGISTER_DECODER (MPC8, mpc8);
     REGISTER_DECODER (NELLYMOSER, nellymoser);
+    REGISTER_DECODER (QCELP, qcelp);
     REGISTER_DECODER (QDM2, qdm2);
     REGISTER_DECODER (RA_144, ra_144);
     REGISTER_DECODER (RA_288, ra_288);
@@ -294,6 +295,7 @@
     REGISTER_PARSER  (MPEGAUDIO, mpegaudio);
     REGISTER_PARSER  (MPEGVIDEO, mpegvideo);
     REGISTER_PARSER  (PNM, pnm);
+    REGISTER_PARSER  (QCELP, qcelp);
     REGISTER_PARSER  (VC1, vc1);
 
     /* bitstream filters */
Index: libavformat/isom.c
===================================================================
--- libavformat/isom.c	(revision 11264)
+++ libavformat/isom.c	(working copy)
@@ -166,6 +166,9 @@
     { CODEC_ID_AC3,  MKTAG('m', 's', 0x20, 0x00) }, /* Dolby AC-3 */
 
     { CODEC_ID_ALAC, MKTAG('a', 'l', 'a', 'c') }, /* Apple Lossless */
+
+    { CODEC_ID_QCELP, MKTAG('Q','c','l','p') },
+
     { CODEC_ID_QDM2, MKTAG('Q', 'D', 'M', '2') }, /* QDM2 */
 
     { CODEC_ID_DVAUDIO, MKTAG('v', 'd', 'v', 'a') },
Index: libavformat/mov.c
===================================================================
--- libavformat/mov.c	(revision 11264)
+++ libavformat/mov.c	(working copy)
@@ -830,6 +830,8 @@
         st->codec->sample_rate= 16000;
         st->codec->channels= 1; /* really needed */
         break;
+    case CODEC_ID_QCELP:
+        st->need_parsing = AVSTREAM_PARSE_FULL;
     case CODEC_ID_AMR_NB:
         st->codec->sample_rate= 8000;
         st->codec->channels= 1; /* really needed */
