Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 9314)
+++ libavcodec/Makefile	(working copy)
@@ -128,6 +128,7 @@
 OBJS-$(CONFIG_PNG_ENCODER)             += png.o
 OBJS-$(CONFIG_PPM_ENCODER)             += pnmenc.o pnm.o
 OBJS-$(CONFIG_PTX_DECODER)             += ptx.o
+OBJS-$(CONFIG_QCELP_DECODER)           += qcelpdec.o
 OBJS-$(CONFIG_QDM2_DECODER)            += qdm2.o mdct.o fft.o mpegaudiodec.o mpegaudiodecheader.o mpegaudio.o mpegaudiodata.o
 OBJS-$(CONFIG_QDRAW_DECODER)           += qdrw.o
 OBJS-$(CONFIG_QPEG_DECODER)            += qpeg.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 9314)
+++ libavcodec/allcodecs.c	(working copy)
@@ -191,6 +191,7 @@
     REGISTER_DECODER(MP3ADU, mp3adu);
     REGISTER_DECODER(MP3ON4, mp3on4);
     REGISTER_DECODER(MPC7, mpc7);
+    REGISTER_DECODER(QCELP, qcelp);
     REGISTER_DECODER(QDM2, qdm2);
     REGISTER_DECODER(RA_144, ra_144);
     REGISTER_DECODER(RA_288, ra_288);
Index: libavcodec/allcodecs.h
===================================================================
--- libavcodec/allcodecs.h	(revision 9314)
+++ libavcodec/allcodecs.h	(working copy)
@@ -144,6 +144,7 @@
 extern AVCodec nuv_decoder;
 extern AVCodec png_decoder;
 extern AVCodec ptx_decoder;
+extern AVCodec qcelp_decoder;
 extern AVCodec qdm2_decoder;
 extern AVCodec qdraw_decoder;
 extern AVCodec qpeg_decoder;
Index: libavformat/mov.c
===================================================================
--- libavformat/mov.c	(revision 9314)
+++ libavformat/mov.c	(working copy)
@@ -832,6 +832,7 @@
         st->codec->sample_rate= 16000;
         st->codec->channels= 1; /* really needed */
         break;
+    case CODEC_ID_QCELP:
     case CODEC_ID_AMR_NB:
         st->codec->sample_rate= 8000;
         st->codec->channels= 1; /* really needed */
