Index: configure
===================================================================
--- configure	(revision 11671)
+++ configure	(working copy)
@@ -72,6 +72,8 @@
   echo "                           and ffmpeg will be unredistributable [default=no]"
   echo "  --enable-pp              enable GPLed postprocessing support [default=no]"
   echo "  --enable-swscaler        software scaler support [default=no]"
+  echo "  --enable-avfilter        video filter support (replaces vhook) [default=no]"
+  echo "  --enable-avfilter-lavf   video filters dependant on avformat [default=no]"
   echo "  --enable-beosthreads     use BeOS threads [default=no]"
   echo "  --enable-os2threads      use OS/2 threads [default=no]"
   echo "  --enable-pthreads        use pthreads [default=no]"
@@ -624,6 +626,8 @@
 
 CONFIG_LIST="
     $COMPONENT_LIST
+    avfilter
+    avfilter_lavf
     avisynth
     beos_netserver
     ffmpeg
@@ -1890,6 +1894,8 @@
 echo "shared                    ${shared-no}"
 echo "postprocessing support    ${pp-no}"
 echo "software scaler enabled   ${swscaler-no}"
+echo "new filter support        ${avfilter-no}"
+echo "filters using lavformat   ${avfilter-lavf-no}"
 echo "video hooking             ${vhook-no}"
 if enabled vhook; then
     echo "Imlib2 support            ${imlib2-no}"
@@ -2087,6 +2093,7 @@
         libavcodec/sh4    \
         libavcodec/sparc  \
         libavdevice       \
+        libavfilter       \
         libavformat       \
         libavutil         \
         libpostproc       \
@@ -2101,6 +2108,7 @@
         doc/texi2pod.pl      \
         libavcodec/Makefile  \
         libavdevice/Makefile \
+        libavfilter/Makefile \
         libavformat/Makefile \
         libavutil/Makefile   \
         libpostproc/Makefile \
@@ -2189,3 +2197,8 @@
     apply libswscale.pc sed s/^Libs:.*$/Libs:/
     apply libswscale-uninstalled.pc sed s/^Libs:.*$/Libs:/
 fi
+
+if enabled avfilter; then
+  pkgconfig_generate libavfilter "FFmpeg video filtering library" "$avfilter_version" "-lavfilter $extralibs" "$pkg_requires libavutil = $lavu=version" ffmpeg
+  pkgconfig_generate_uninstalled libavfilter "FFmpeg video filtering library" "$avfilter_version" "$extralibs" "$pkg_requires libavutil = $libavu_version"
+fi
Index: Makefile
===================================================================
--- Makefile	(revision 11671)
+++ Makefile	(working copy)
@@ -8,7 +8,7 @@
 
 CFLAGS=$(OPTFLAGS) -I$(BUILD_ROOT) -I$(SRC_PATH) -I$(SRC_PATH)/libavutil \
        -I$(SRC_PATH)/libavcodec -I$(SRC_PATH)/libavformat -I$(SRC_PATH)/libswscale \
-       -I$(SRC_PATH)/libavdevice \
+       -I$(SRC_PATH)/libavdevice -I$(SRC_PATH)/libavfilter \
        -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_ISOC9X_SOURCE -DHAVE_AV_CONFIG_H
 LDFLAGS+= -g
 
@@ -69,6 +69,11 @@
 EXTRALIBS+=-lswscale$(BUILDSUF)
 endif
 
+ifeq ($(CONFIG_AVFILTER),yes)
+LDFLAGS+=-L$(BUILD_ROOT)/libavfilter
+EXTRALIBS := -lavfilter$(BUILDSUF) $(EXTRALIBS)
+endif
+
 MAKE-yes = $(MAKE)
 MAKE-    = : $(MAKE)
 
@@ -81,6 +86,7 @@
 	$(MAKE)                    -C libavdevice all
 	$(MAKE-$(CONFIG_PP))       -C libpostproc all
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  all
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter all
 
 ffmpeg_g$(EXESUF): ffmpeg.o cmdutils.o .libs
 	$(CC) $(LDFLAGS) -o $@ ffmpeg.o cmdutils.o $(EXTRALIBS)
@@ -166,6 +172,7 @@
 	$(MAKE)                    -C libavdevice install-libs
 	$(MAKE-$(CONFIG_PP))       -C libpostproc install-libs
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  install-libs
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter install-libs
 ifeq ($(BUILD_SHARED),yes)
 	-$(LDCONFIG)
 endif
@@ -177,6 +184,7 @@
 	$(MAKE)              -C libavdevice install-headers
 	$(MAKE-$(CONFIG_PP)) -C libpostproc install-headers
 	$(MAKE)              -C libswscale  install-headers
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter install-headers
 
 uninstall: uninstall-progs uninstall-libs uninstall-headers uninstall-man uninstall-vhook
 
@@ -197,6 +205,7 @@
 	$(MAKE) -C libavdevice uninstall-libs
 	$(MAKE) -C libpostproc uninstall-libs
 	$(MAKE) -C libswscale  uninstall-libs
+	$(MAKE) -C libavfilter uninstall-libs
 
 uninstall-headers:
 	$(MAKE) -C libavutil   uninstall-headers
@@ -205,6 +214,7 @@
 	$(MAKE) -C libavdevice uninstall-headers
 	$(MAKE) -C libpostproc uninstall-headers
 	$(MAKE) -C libswscale  uninstall-headers
+	$(MAKE) -C libavfilter uninstall-headers
 	-rmdir "$(INCDIR)"
 
 depend dep: .depend .vhookdep
@@ -214,6 +224,7 @@
 	$(MAKE)                    -C libavdevice depend
 	$(MAKE-$(CONFIG_PP))       -C libpostproc depend
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  depend
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter depend
 
 .depend: $(SRCS) version.h
 	$(CC) -MM $(CFLAGS) $(SDL_CFLAGS) $(filter-out %.h,$^) 1>.depend
@@ -234,6 +245,7 @@
 	$(MAKE) -C libavdevice clean
 	$(MAKE) -C libpostproc clean
 	$(MAKE) -C libswscale  clean
+	$(MAKE) -C libavfilter clean
 	rm -f *.o *~ .libs gmon.out TAGS $(ALLPROGS) $(ALLPROGS_G) \
 	   output_example$(EXESUF)
 	rm -f doc/*.html doc/*.pod doc/*.1
@@ -249,6 +261,7 @@
 	$(MAKE) -C libavdevice distclean
 	$(MAKE) -C libpostproc distclean
 	$(MAKE) -C libswscale  distclean
+	$(MAKE) -C libavfilter distclean
 	rm -f .depend .vhookdep version.h config.* *.pc
 
 TAGS:
