Index: Makefile
===================================================================
--- Makefile	(revision 11882)
+++ Makefile	(working copy)
@@ -8,7 +8,7 @@
 
 CFLAGS=$(OPTFLAGS) -I$(BUILD_ROOT) -I$(SRC_PATH) -I$(SRC_PATH)/libavutil \
        -I$(SRC_PATH)/libavcodec -I$(SRC_PATH)/libavformat -I$(SRC_PATH)/libswscale \
-       -I$(SRC_PATH)/libavdevice \
+       -I$(SRC_PATH)/libavdevice -I$(SRC_PATH)/libavfilter \
        -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_ISOC9X_SOURCE -DHAVE_AV_CONFIG_H
 LDFLAGS+= -g
 
@@ -69,6 +69,11 @@
 EXTRALIBS+=-lswscale$(BUILDSUF)
 endif
 
+ifeq ($(CONFIG_AVFILTER),yes)
+LDFLAGS+=-L$(BUILD_ROOT)/libavfilter
+EXTRALIBS := -lavfilter$(BUILDSUF) $(EXTRALIBS)
+endif
+
 MAKE-yes = $(MAKE)
 MAKE-    = : $(MAKE)
 
@@ -81,6 +86,7 @@
 	$(MAKE)                    -C libavdevice all
 	$(MAKE-$(CONFIG_PP))       -C libpostproc all
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  all
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter all
 
 ffmpeg_g$(EXESUF): ffmpeg.o cmdutils.o .libs
 	$(CC) $(LDFLAGS) -o $@ ffmpeg.o cmdutils.o $(EXTRALIBS)
@@ -166,6 +172,7 @@
 	$(MAKE)                    -C libavdevice install-libs
 	$(MAKE-$(CONFIG_PP))       -C libpostproc install-libs
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  install-libs
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter install-libs
 ifeq ($(BUILD_SHARED),yes)
 	-$(LDCONFIG)
 endif
@@ -177,6 +184,7 @@
 	$(MAKE)              -C libavdevice install-headers
 	$(MAKE-$(CONFIG_PP)) -C libpostproc install-headers
 	$(MAKE)              -C libswscale  install-headers
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter install-headers
 
 uninstall: uninstall-progs uninstall-libs uninstall-headers uninstall-man uninstall-vhook
 
@@ -197,6 +205,7 @@
 	$(MAKE) -C libavdevice uninstall-libs
 	$(MAKE) -C libpostproc uninstall-libs
 	$(MAKE) -C libswscale  uninstall-libs
+	$(MAKE) -C libavfilter uninstall-libs
 
 uninstall-headers:
 	$(MAKE) -C libavutil   uninstall-headers
@@ -205,6 +214,7 @@
 	$(MAKE) -C libavdevice uninstall-headers
 	$(MAKE) -C libpostproc uninstall-headers
 	$(MAKE) -C libswscale  uninstall-headers
+	$(MAKE) -C libavfilter uninstall-headers
 	-rmdir "$(INCDIR)"
 
 depend dep: .depend .vhookdep
@@ -214,6 +224,7 @@
 	$(MAKE)                    -C libavdevice depend
 	$(MAKE-$(CONFIG_PP))       -C libpostproc depend
 	$(MAKE-$(CONFIG_SWSCALER)) -C libswscale  depend
+	$(MAKE-$(CONFIG_AVFILTER)) -C libavfilter depend
 
 .depend: $(SRCS) version.h
 	$(CC) -MM $(CFLAGS) $(SDL_CFLAGS) $(filter-out %.h,$^) 1>.depend
@@ -234,6 +245,7 @@
 	$(MAKE) -C libavdevice clean
 	$(MAKE) -C libpostproc clean
 	$(MAKE) -C libswscale  clean
+	$(MAKE) -C libavfilter clean
 	rm -f *.o *~ .libs gmon.out TAGS $(ALLPROGS) $(ALLPROGS_G) \
 	   output_example$(EXESUF)
 	rm -f doc/*.html doc/*.pod doc/*.1
@@ -249,6 +261,7 @@
 	$(MAKE) -C libavdevice distclean
 	$(MAKE) -C libpostproc distclean
 	$(MAKE) -C libswscale  distclean
+	$(MAKE) -C libavfilter distclean
 	rm -f .depend .vhookdep version.h config.* *.pc
 
 TAGS:
Index: configure
===================================================================
--- configure	(revision 11882)
+++ configure	(working copy)
@@ -72,6 +72,8 @@
   echo "                           and ffmpeg will be unredistributable [default=no]"
   echo "  --enable-pp              enable GPLed postprocessing support [default=no]"
   echo "  --enable-swscaler        software scaler support [default=no]"
+  echo "  --enable-avfilter        video filter support (replaces vhook) [default=no]"
+  echo "  --enable-avfilter-lavf   video filters dependant on avformat [default=no]"
   echo "  --enable-beosthreads     use BeOS threads [default=no]"
   echo "  --enable-os2threads      use OS/2 threads [default=no]"
   echo "  --enable-pthreads        use pthreads [default=no]"
@@ -620,10 +622,13 @@
     outdevs
     parsers
     protocols
+    filters
 "
 
 CONFIG_LIST="
     $COMPONENT_LIST
+    avfilter
+    avfilter_lavf
     avisynth
     beos_netserver
     ffmpeg
@@ -959,6 +964,9 @@
 OUTDEV_LIST=`sed -n 's/^[^#]*_MUX.*(.*, *\(.*\)).*/\1_muxer/p' "$source_path/libavdevice/alldevices.c"`
 INDEV_LIST=`sed -n 's/^[^#]*DEMUX.*(.*, *\(.*\)).*/\1_demuxer/p' "$source_path/libavdevice/alldevices.c"`
 PROTOCOL_LIST=`sed -n 's/^[^#]*PROTOCOL.*(.*, *\(.*\)).*/\1_protocol/p' "$source_path/libavformat/allformats.c"`
+VIDEO_FILTER_LIST=`sed -n 's/^[^#]*_VF.*(.*, *\(.*\)).*/vf_\1/p' "$source_path/libavfilter/allfilters.c"`
+SRC_FILTER_LIST=`sed -n 's/^[^#]*_VSRC.*(.*, *\(.*\)).*/vsrc_\1/p' "$source_path/libavfilter/allfilters.c"`
+FILTER_LIST="$SRC_FILTER_LIST $VIDEO_FILTER_LIST"
 
 enable $ARCH_EXT_LIST \
        $DECODER_LIST \
@@ -967,6 +975,7 @@
        $BSF_LIST \
        $DEMUXER_LIST \
        $MUXER_LIST \
+       $FILTER_LIST \
        $PROTOCOL_LIST \
        $INDEV_LIST \
        $OUTDEV_LIST \
@@ -1035,6 +1044,8 @@
     ;;
     --disable-muxers) disable $MUXER_LIST
     ;;
+    --disable-filters) disable $FILTER_LIST
+    ;;
     --disable-demuxers) disable $DEMUXER_LIST
     ;;
     --disable-parsers) disable $PARSER_LIST
@@ -1048,7 +1059,8 @@
     --enable-*=*|--disable-*=*)
     eval `echo "$opt" | sed 's/=/-/;s/--/action=/;s/-/ thing=/;s/-/ name=/'`
     case "$thing" in
-        encoder|decoder|muxer|demuxer|parser|bsf|protocol) $action ${optval}_${thing} ;;
+        encoder|decoder|muxer|demuxer|parser|bsf|protocol) $action ${optval}_${thing} ;;	
+	filter) $action ${optval} ;;	
         *) die_unknown "$opt" ;;
     esac
     ;;
@@ -1834,6 +1846,7 @@
 enabled_any $BSF_LIST          && enable bsfs
 enabled_any $DEMUXER_LIST      && enable demuxers
 enabled_any $MUXER_LIST        && enable muxers
+enabled_any $FILTER_LIST       && enable filters
 enabled_any $INDEV_LIST        && enable demuxers
 enabled_any $OUTDEV_LIST       && enable muxers
 enabled_any $PROTOCOL_LIST     && enable protocols
@@ -1848,6 +1861,7 @@
            $BSF_LIST          \
            $DEMUXER_LIST      \
            $MUXER_LIST        \
+           $FILTER_LIST       \
            $INDEV_LIST        \
            $OUTDEV_LIST       \
            $PROTOCOL_LIST     \
@@ -1893,6 +1907,8 @@
 echo "shared                    ${shared-no}"
 echo "postprocessing support    ${pp-no}"
 echo "software scaler enabled   ${swscaler-no}"
+echo "new filter support        ${avfilter-no}"
+echo "filters using lavformat   ${avfilter-lavf-no}"
 echo "video hooking             ${vhook-no}"
 if enabled vhook; then
     echo "Imlib2 support            ${imlib2-no}"
@@ -1926,7 +1942,7 @@
 echo "XviD enabled              ${libxvid-no}"
 echo "zlib enabled              ${zlib-no}"
 
-for type in decoder encoder parser demuxer muxer protocol bsf indev outdev; do
+for type in decoder encoder parser demuxer muxer protocol filter bsf indev outdev; do
     echo "Enabled ${type}s:"
     ucname="\$`toupper $type`_LIST"
     list="`eval echo $ucname`"
@@ -2009,6 +2025,7 @@
 lavd_version=`grep '#define LIBAVDEVICE_VERSION ' "$source_path/libavdevice/avdevice.h" | sed 's/[^0-9\.]//g'`
 lavf_version=`grep '#define LIBAVFORMAT_VERSION ' "$source_path/libavformat/avformat.h" | sed 's/[^0-9\.]//g'`
 lavu_version=`grep '#define LIBAVUTIL_VERSION ' "$source_path/libavutil/avutil.h" | sed 's/[^0-9\.]//g'`
+lavfi_version=`grep '#define LIBAVFILTER_VERSION ' "$source_path/libavfilter/avfilter.h" | sed 's/[^0-9\.]//g'`
 
 
 
@@ -2025,6 +2042,8 @@
     echo "LAVFVERSION=$lavf_version" >> config.mak
     echo "LAVUMAJOR=${lavu_version%%.*}" >> config.mak
     echo "LAVUVERSION=$lavu_version" >> config.mak
+    echo "LAVFILTERMAJOR=${lavfi_version%%.*}" >> config.mak
+    echo "LAVFILTERVERSION=$lavfi_version" >> config.mak
     echo "SWSMAJOR=${sws_version%%.*}" >> config.mak
     echo "SWSVERSION=$sws_version" >> config.mak
     echo "SLIBNAME=${SLIBNAME}" >> config.mak
@@ -2047,6 +2066,7 @@
                                       $BSF_LIST          \
                                       $DEMUXER_LIST      \
                                       $MUXER_LIST        \
+                                      $FILTER_LIST       \
                                       $PROTOCOL_LIST     \
                                       $INDEV_LIST        \
                                       $OUTDEV_LIST       \
@@ -2090,6 +2110,7 @@
         libavcodec/sh4    \
         libavcodec/sparc  \
         libavdevice       \
+        libavfilter       \
         libavformat       \
         libavutil         \
         libpostproc       \
@@ -2104,6 +2125,7 @@
         doc/texi2pod.pl      \
         libavcodec/Makefile  \
         libavdevice/Makefile \
+        libavfilter/Makefile \
         libavformat/Makefile \
         libavutil/Makefile   \
         libpostproc/Makefile \
@@ -2192,3 +2214,8 @@
     apply libswscale.pc sed s/^Libs:.*$/Libs:/
     apply libswscale-uninstalled.pc sed s/^Libs:.*$/Libs:/
 fi
+
+if enabled avfilter; then
+  pkgconfig_generate libavfilter "FFmpeg video filtering library" "$avfilter_version" "-lavfilter $extralibs" "$pkg_requires libavutil = $lavu=version" ffmpeg
+  pkgconfig_generate_uninstalled libavfilter "FFmpeg video filtering library" "$avfilter_version" "$extralibs" "$pkg_requires libavutil = $libavu_version"
+fi
