Index: configure
===================================================================
--- configure	(revision 23145)
+++ configure	(working copy)
@@ -172,6 +172,7 @@
   --enable-libfaac         enable FAAC support via libfaac [no]
   --enable-libfaad         enable FAAD support via libfaad [no]
   --enable-libfaadbin      open libfaad.so.0 at runtime [no]
+  --enable-libfreetype     enable libfreetype [no]
   --enable-libgsm          enable GSM support via libgsm [no]
   --enable-libmp3lame      enable MP3 encoding via libmp3lame [no]
   --enable-libnut          enable NUT (de)muxing via libnut,
@@ -916,6 +917,7 @@
     libfaac
     libfaad
     libfaadbin
+    libfreetype
     libgsm
     libmp3lame
     libnut
@@ -1056,6 +1058,7 @@
     llrintf
     local_aligned_16
     local_aligned_8
+    localtime_r
     log2
     log2f
     loongson
@@ -1413,6 +1416,7 @@
 udp_protocol_deps="network"
 
 # filters
+drawtext_filter_deps="libfreetype"
 movie_filter_deps="avfilter_lavf"
 avfilter_lavf_deps="avformat"
 
@@ -2523,6 +2527,7 @@
 check_struct "sys/time.h sys/resource.h" "struct rusage" ru_maxrss
 check_func  inet_aton $network_extralibs
 check_func  isatty
+check_func  localtime_r
 check_func  ${malloc_prefix}memalign            && enable memalign
 check_func  mkstemp
 check_func  ${malloc_prefix}posix_memalign      && enable posix_memalign
@@ -2605,6 +2610,7 @@
                       require  libdirac libdirac_encoder/dirac_encoder.h dirac_encoder_init $(pkg-config --libs dirac)
 enabled libfaac    && require2 libfaac "stdint.h faac.h" faacEncGetVersion -lfaac
 enabled libfaad    && require2 libfaad faad.h faacDecOpen -lfaad
+enabled libfreetype && add_cflags $(pkg-config --cflags freetype2) && require libfreetype ft2build.h FT_Init_FreeType -lfreetype -lz
 enabled libgsm     && require  libgsm gsm/gsm.h gsm_create -lgsm
 enabled libmp3lame && require  libmp3lame lame/lame.h lame_init -lmp3lame -lm
 enabled libnut     && require  libnut libnut.h nut_demuxer_init -lnut
