--- ffplay.c.old	2008-02-29 21:39:16.000000000 +0100
+++ ffplay.c	2008-02-29 21:40:36.000000000 +0100
@@ -27,6 +27,11 @@
 #include "swscale.h"
 #include "avstring.h"
 
+#if ENABLE_AVFILTER
+# include "avfilter.h"
+# include "avfiltergraph.h"
+#endif
+
 #include "version.h"
 #include "cmdutils.h"
 
@@ -66,7 +71,9 @@
 /* NOTE: the size must be big enough to compensate the hardware audio buffersize size */
 #define SAMPLE_ARRAY_SIZE (2*65536)
 
+#if !ENABLE_AVFILTER
 static int sws_flags = SWS_BICUBIC;
+#endif
 
 typedef struct PacketQueue {
     AVPacketList *first_pkt, *last_pkt;
@@ -86,6 +93,10 @@
     int width, height; /* source height & width */
     enum PixelFormat pix_fmt;
     int allocated;
+
+#if ENABLE_AVFILTER
+    AVFilterPicRef *picref;
+#endif
 } VideoPicture;
 
 typedef struct SubPicture {
@@ -168,6 +179,10 @@
     //    QETimer *video_timer;
     char filename[1024];
     int width, height, xleft, ytop;
+
+#if ENABLE_AVFILTER
+    AVFilterContext *out_video_filter;          ///<the last filter in the video chain
+#endif
 } VideoState;
 
 void show_help(void);
@@ -207,6 +222,9 @@
 static int error_resilience = FF_ER_CAREFUL;
 static int error_concealment = 3;
 static int decoder_reorder_pts= 0;
+#if ENABLE_AVFILTER
+static char *vfilters = NULL;
+#endif
 
 /* current context */
 static int is_full_screen;
@@ -653,14 +671,22 @@
 
     vp = &is->pictq[is->pictq_rindex];
     if (vp->bmp) {
+#if ENABLE_AVFILTER
+        if (vp->picref->pixel_aspect.num == 0)
+            aspect_ratio = 0;
+        else
+            aspect_ratio = av_q2d(vp->picref->pixel_aspect);
+#else
         /* XXX: use variable in the frame */
         if (is->video_st->codec->sample_aspect_ratio.num == 0)
             aspect_ratio = 0;
         else
-            aspect_ratio = av_q2d(is->video_st->codec->sample_aspect_ratio)
-                * vp->width / vp->height;
+            aspect_ratio = av_q2d(is->video_st->codec->sample_aspect_ratio);
+#endif
         if (aspect_ratio <= 0.0)
             aspect_ratio = (float)vp->width / (float)vp->height;
+        else
+            aspect_ratio *= (float)vp->width / (float)vp->height;
         /* if an active format is indicated, then it overrides the
            mpeg format */
 #if 0
@@ -863,9 +889,15 @@
     } else if(!is_full_screen && screen_width){
         w = screen_width;
         h = screen_height;
+#if ENABLE_AVFILTER
+    }else if (is->out_video_filter && is->out_video_filter->inputs[0]){
+        w = is->out_video_filter->inputs[0]->w;
+        h = is->out_video_filter->inputs[0]->h;
+#else
     }else if (is->video_st && is->video_st->codec->width){
         w = is->video_st->codec->width;
         h = is->video_st->codec->height;
+#endif
     } else {
         w = 640;
         h = 480;
@@ -1171,9 +1203,19 @@
     if (vp->bmp)
         SDL_FreeYUVOverlay(vp->bmp);
 
+#if ENABLE_AVFILTER
+    if (vp->picref)
+        avfilter_unref_pic(vp->picref);
+    vp->picref = NULL;
+
+    vp->width   = is->out_video_filter->inputs[0]->w;
+    vp->height  = is->out_video_filter->inputs[0]->h;
+    vp->pix_fmt = is->out_video_filter->inputs[0]->format;
+#else
     vp->width   = is->video_st->codec->width;
     vp->height  = is->video_st->codec->height;
     vp->pix_fmt = is->video_st->codec->pix_fmt;
+#endif
 
 #if 0
     /* XXX: use generic function */
@@ -1211,8 +1253,11 @@
     VideoPicture *vp;
     int dst_pix_fmt;
     AVPicture pict;
+#if ENABLE_AVFILTER
+    AVPicture pict_src;
+#else
     static struct SwsContext *img_convert_ctx;
-
+#endif
     /* wait until we have space to put a new picture */
     SDL_LockMutex(is->pictq_mutex);
     while (is->pictq_size >= VIDEO_PICTURE_QUEUE_SIZE &&
@@ -1228,8 +1273,13 @@
 
     /* alloc or resize hardware picture buffer */
     if (!vp->bmp ||
+#if ENABLE_AVFILTER
+        vp->width  != is->out_video_filter->inputs[0]->w ||
+        vp->height != is->out_video_filter->inputs[0]->h) {
+#else
         vp->width != is->video_st->codec->width ||
         vp->height != is->video_st->codec->height) {
+#endif
         SDL_Event event;
 
         vp->allocated = 0;
@@ -1253,6 +1303,12 @@
 
     /* if the frame is not skipped, then display it */
     if (vp->bmp) {
+#if ENABLE_AVFILTER
+        if(vp->picref)
+            avfilter_unref_pic(vp->picref);
+        vp->picref = src_frame->opaque;
+#endif
+
         /* get a pointer on the bitmap */
         SDL_LockYUVOverlay (vp->bmp);
 
@@ -1264,6 +1320,19 @@
         pict.linesize[0] = vp->bmp->pitches[0];
         pict.linesize[1] = vp->bmp->pitches[2];
         pict.linesize[2] = vp->bmp->pitches[1];
+
+#if ENABLE_AVFILTER
+        pict_src.data[0] = src_frame->data[0];
+        pict_src.data[1] = src_frame->data[1];
+        pict_src.data[2] = src_frame->data[2];
+
+        pict_src.linesize[0] = src_frame->linesize[0];
+        pict_src.linesize[1] = src_frame->linesize[1];
+        pict_src.linesize[2] = src_frame->linesize[2];
+
+        av_picture_copy(&pict, &pict_src,
+                        vp->pix_fmt, vp->width, vp->height);
+#else
         img_convert_ctx = sws_getCachedContext(img_convert_ctx,
             vp->width, vp->height, vp->pix_fmt, vp->width, vp->height,
             dst_pix_fmt, sws_flags, NULL, NULL, NULL);
@@ -1273,6 +1342,7 @@
         }
         sws_scale(img_convert_ctx, src_frame->data, src_frame->linesize,
                   0, vp->height, pict.data, pict.linesize);
+#endif
         /* update the bitmap content */
         SDL_UnlockYUVOverlay(vp->bmp);
 
@@ -1382,6 +1452,138 @@
     return 0;
 }
 
+#if ENABLE_AVFILTER
+typedef struct {
+    VideoState *is;
+    AVFrame *frame;
+} FilterPriv;
+
+static int input_init(AVFilterContext *ctx, const char *args, void *opaque)
+{
+    FilterPriv *priv = ctx->priv;
+    if(!opaque) return -1;
+
+    priv->is = opaque;
+    priv->frame = avcodec_alloc_frame();
+
+    return 0;
+}
+
+static void input_uninit(AVFilterContext *ctx)
+{
+    FilterPriv *priv = ctx->priv;
+    av_free(priv->frame);
+}
+
+static int input_request_frame(AVFilterLink *link)
+{
+    FilterPriv *priv = link->src->priv;
+    AVFilterPicRef *picref;
+    uint64_t pts;
+
+    if(get_video_frame(priv->is, priv->frame, &pts) < 0)
+        return -1;
+
+    /* FIXME: until I figure out how to hook everything up to the codec
+     * right, we're just copying the entire frame. */
+    picref = avfilter_get_video_buffer(link, AV_PERM_WRITE);
+    av_picture_copy((AVPicture *)&picref->data, (AVPicture *)priv->frame,
+                    picref->pic->format, picref->w, picref->h);
+
+    picref->pts = pts;
+    picref->pixel_aspect = priv->is->video_st->codec->sample_aspect_ratio;
+    avfilter_start_frame(link, avfilter_ref_pic(picref, ~0));
+    avfilter_draw_slice(link, 0, picref->h);
+    avfilter_end_frame(link);
+    avfilter_unref_pic(picref);
+
+    return 0;
+}
+
+static int input_query_formats(AVFilterContext *ctx)
+{
+    FilterPriv *priv = ctx->priv;
+    avfilter_set_common_formats(ctx,
+            avfilter_make_format_list(1, priv->is->video_st->codec->pix_fmt));
+    return 0;
+}
+
+static int input_config_props(AVFilterLink *link)
+{
+    FilterPriv *priv  = link->src->priv;
+    AVCodecContext *c = priv->is->video_st->codec;
+
+    link->w = c->width;
+    link->h = c->height;
+
+    return 0;
+}
+
+static AVFilter input_filter =
+{
+    .name      = "ffplay_input",
+
+    .priv_size = sizeof(FilterPriv),
+
+    .init      = input_init,
+    .uninit    = input_uninit,
+
+    .query_formats = input_query_formats,
+
+    .inputs    = (AVFilterPad[]) {{ .name = NULL }},
+    .outputs   = (AVFilterPad[]) {{ .name = "default",
+                                    .type = CODEC_TYPE_VIDEO,
+                                    .request_frame = input_request_frame,
+                                    .config_props  = input_config_props, },
+                                  { .name = NULL }},
+};
+
+static void output_end_frame(AVFilterLink *link)
+{
+}
+
+static int output_query_formats(AVFilterContext *ctx)
+{
+    avfilter_set_common_formats(ctx,
+        avfilter_make_format_list(1, PIX_FMT_YUV420P));
+    return 0;
+}
+
+static int get_filtered_video_frame(AVFilterContext *ctx, AVFrame *frame,
+                                    uint64_t *pts)
+{
+    AVFilterPicRef *pic;
+
+    if(avfilter_request_frame(ctx->inputs[0]))
+        return -1;
+    if(!(pic = ctx->inputs[0]->cur_pic))
+        return -1;
+    ctx->inputs[0]->cur_pic = NULL;
+
+    frame->opaque = pic;
+    *pts          = pic->pts;
+
+    memcpy(frame->data,     pic->data,     sizeof(frame->data));
+    memcpy(frame->linesize, pic->linesize, sizeof(frame->linesize));
+
+    return 1;
+}
+
+static AVFilter output_filter =
+{
+    .name      = "ffplay_output",
+
+    .query_formats = output_query_formats,
+
+    .inputs    = (AVFilterPad[]) {{ .name          = "default",
+                                    .type          = CODEC_TYPE_VIDEO,
+                                    .end_frame     = output_end_frame,
+                                    .min_perms     = AV_PERM_READ, },
+                                  { .name = NULL }},
+    .outputs   = (AVFilterPad[]) {{ .name = NULL }},
+};
+#endif  /* ENABLE_AVFILTER */
+
 static int video_thread(void *arg)
 {
     VideoState *is = arg;
@@ -1390,10 +1592,52 @@
     double pts;
     int ret;
 
+
+#if ENABLE_AVFILTER
+    AVFilterContext *filt_src = NULL, *filt_out = NULL, *filt_graph = NULL;
+    AVFilterContext *graph = NULL;
+
+    avfilter_register_all();
+
+    if(!(filt_src = avfilter_open(&input_filter,  "src")))   goto the_end;
+    if(!(filt_out = avfilter_open(&output_filter, "out")))   goto the_end;
+
+    if(avfilter_init_filter(filt_src, NULL, is))             goto the_end;
+    if(avfilter_init_filter(filt_out, NULL, frame))          goto the_end;
+
+    if(!(graph = avfilter_open(avfilter_get_by_name("graph"), NULL)))
+        goto the_end;
+    if(avfilter_init_filter(graph, NULL, NULL))
+        goto the_end;
+
+    if(vfilters) {
+        if(!(filt_graph = avfilter_open(avfilter_get_by_name("graph"), NULL)))
+            goto the_end;
+        if(avfilter_init_filter(filt_graph, vfilters, NULL))
+            goto the_end;
+        if(avfilter_link(filt_src, 0, filt_graph, 0))        goto the_end;
+        if(avfilter_link(filt_graph, 0, filt_out, 0))        goto the_end;
+
+        avfilter_graph_add_filter(graph, filt_graph);
+    } else {
+    if(avfilter_link(filt_src, 0, filt_out, 0) < 0)          goto the_end;
+    }
+    avfilter_graph_add_filter(graph, filt_src);
+    avfilter_graph_add_filter(graph, filt_out);
+    if(avfilter_graph_config_formats(graph))                 goto the_end;
+    if(avfilter_config_links(filt_out))                      goto the_end;
+
+    is->out_video_filter = filt_out;
+#endif
+
     for(;;) {
         while (is->paused && !is->videoq.abort_request)
             SDL_Delay(10);
+#if ENABLE_AVFILTER
+        ret = get_filtered_video_frame(filt_out, frame, &pts_int);
+#else
         ret = get_video_frame(is, frame, &pts_int);
+#endif
 
         if (ret < 0) goto the_end;
 
@@ -1410,6 +1654,9 @@
                 stream_pause(cur_stream);
     }
  the_end:
+ #if ENABLE_AVFILTER
+    if(graph) avfilter_destroy(graph);
+ #endif
     av_free(frame);
     return 0;
 }
@@ -2149,6 +2396,12 @@
     /* free all pictures */
     for(i=0;i<VIDEO_PICTURE_QUEUE_SIZE; i++) {
         vp = &is->pictq[i];
+#if ENABLE_AVFILTER
+        if (vp->picref) {
+            avfilter_unref_pic(vp->picref);
+            vp->picref = NULL;
+        }
+#endif
         if (vp->bmp) {
             SDL_FreeYUVOverlay(vp->bmp);
             vp->bmp = NULL;
@@ -2494,6 +2747,9 @@
     { "ec", OPT_INT | HAS_ARG | OPT_EXPERT, {(void*)&error_concealment}, "set error concealment options",  "bit_mask" },
     { "sync", HAS_ARG | OPT_EXPERT, {(void*)opt_sync}, "set audio-video sync. type (type=audio/video/ext)", "type" },
     { "threads", HAS_ARG | OPT_EXPERT, {(void*)opt_thread_count}, "thread count", "count" },
+#if ENABLE_AVFILTER
+    { "vfilters", OPT_STRING | HAS_ARG, {(void*)&vfilters}, "video filters", "filter list" },
+#endif
     { NULL, },
 };
 
