diff --git a/libavformat/Makefile b/libavformat/Makefile
index 8e8a869..d2aaf47 100644
--- a/libavformat/Makefile
+++ b/libavformat/Makefile
@@ -34,6 +34,7 @@ OBJS-$(CONFIG_BETHSOFTVID_DEMUXER)       += bethsoftvid.o
 OBJS-$(CONFIG_BFI_DEMUXER)               += bfi.o
 OBJS-$(CONFIG_C93_DEMUXER)               += c93.o vocdec.o voc.o
 OBJS-$(CONFIG_CAVSVIDEO_DEMUXER)         += raw.o
+OBJS-$(CONFIG_CONCAT_DEMUXER)            += m3u.o pls.o playlist.o concat.o concatgen.o datanode.o
 OBJS-$(CONFIG_CRC_MUXER)                 += crcenc.o
 OBJS-$(CONFIG_DAUD_DEMUXER)              += daud.o
 OBJS-$(CONFIG_DAUD_MUXER)                += daud.o
