diff --git a/libavformat/utils.c b/libavformat/utils.c
index 0c1a50d..be7a0fd 100644
--- a/libavformat/utils.c
+++ b/libavformat/utils.c
@@ -27,6 +27,7 @@
 #include <sys/time.h>
 #include <time.h>
 #include <strings.h>
+#include "concat.h"
 
 #undef NDEBUG
 #include <assert.h>
@@ -433,6 +434,7 @@ int av_open_input_file(AVFormatContext **ic_ptr, const char *filename,
                        AVFormatParameters *ap)
 {
     int err, probe_size;
+    PlaylistContext *playlist_ctx;
     AVProbeData probe_data, *pd = &probe_data;
     ByteIOContext *pb = NULL;
 
@@ -442,6 +444,17 @@ int av_open_input_file(AVFormatContext **ic_ptr, const char *filename,
     pd->buf = NULL;
     pd->buf_size = 0;
 
+    playlist_ctx = ff_playlist_from_encodedstring(filename, ',');
+    if (playlist_ctx) {
+        av_log((*ic_ptr), AV_LOG_DEBUG, "Generating playlist from %s\n", filename);
+        av_strlcpy((*ic_ptr)->filename, filename, sizeof((*ic_ptr)->filename));
+        (*ic_ptr)->iformat = ff_concat_alloc_demuxer();
+        ff_playlist_set_context((*ic_ptr), playlist_ctx);
+        ff_playlist_populate_context(playlist_ctx, playlist_ctx->pe_curidx);
+        ff_playlist_set_streams((*ic_ptr));
+        return 0;
+    }
+
     if (!fmt) {
         /* guess format if no file can be opened */
         fmt = av_probe_input_format(pd, 0);
@@ -910,6 +923,10 @@ static int av_read_frame_internal(AVFormatContext *s, AVPacket *pkt)
 {
     AVStream *st;
     int len, ret, i;
+    int stream_index;
+    AVStream *stream;
+    stream_index = 0;
+    stream = 0;
 
     av_init_packet(pkt);
 
@@ -943,8 +960,14 @@ static int av_read_frame_internal(AVFormatContext *s, AVPacket *pkt)
                 /* return packet if any */
                 if (pkt->size) {
                 got_packet:
+                    if (stream && stream->codec && stream->codec->codec) {
+                        pkt->stream = stream;
+                        pkt->stream_index = stream_index;
+                    } else {
+                        pkt->stream = st;
+                        pkt->stream_index = st->index;
+                    }
                     pkt->duration = 0;
-                    pkt->stream_index = st->index;
                     pkt->pts = st->parser->pts;
                     pkt->dts = st->parser->dts;
                     pkt->pos = st->parser->pos;
@@ -980,8 +1003,11 @@ static int av_read_frame_internal(AVFormatContext *s, AVPacket *pkt)
                                         NULL, 0,
                                         AV_NOPTS_VALUE, AV_NOPTS_VALUE,
                                         AV_NOPTS_VALUE);
-                        if (pkt->size)
+                        if (pkt->size) {
+                            stream_index = cur_pkt.stream_index;
+                            stream = cur_pkt.stream;
                             goto got_packet;
+                        }
                     }
                 }
                 /* no more packets: really terminate parsing */
