Index: configure
===================================================================
--- configure	(revision 7261)
+++ configure	(working copy)
@@ -79,10 +79,6 @@
   echo "  --disable-static         do not build static libraries [default=no]"
   echo "  --enable-shared          build shared libraries [default=no]"
   echo "  --disable-shared         do not build shared libraries [default=yes]"
-  echo "  --enable-amr_nb          enable amr_nb float audio codec"
-  echo "  --enable-amr_nb-fixed    use fixed point for amr-nb codec"
-  echo "  --enable-amr_wb          enable amr_wb float audio codec"
-  echo "  --enable-amr_if2         enable amr_wb IF2 audio codec"
   echo "  --enable-sunmlib         use Sun medialib [default=no]"
   echo "  --enable-pthreads        use pthreads [default=no]"
   echo "  --enable-dc1394          enable IIDC-1394 grabbing using libdc1394"
@@ -398,10 +394,7 @@
     demuxers
     a52
     a52bin
-    amr
     amr_nb
-    amr_nb_fixed
-    amr_wb
     audio_beos
     audio_oss
     avisynth
@@ -623,10 +616,7 @@
 SLIBNAME_WITH_MAJOR='$(SLIBNAME).$(LIBMAJOR)'
 EXESUF=""
 BUILDSUF=""
-amr_nb="no"
-amr_wb="no"
-amr_nb_fixed="no"
-amr_if2="no"
+amr_nb="yes"
 mlib="no"
 pthreads="no"
 swscaler="no"
@@ -974,14 +964,6 @@
   ;;
   --enable-small) optimize="small"
   ;;
-  --enable-amr_nb) amr="yes"; amr_nb="yes"; amr_nb_fixed="no"
-  ;;
-  --enable-amr_nb-fixed) amr="yes"; amr_nb_fixed="yes"; amr_nb="no"
-  ;;
-  --enable-amr_wb) amr="yes"; amr_wb="yes"
-  ;;
-  --enable-amr_if2) amr="yes"; amr_if2="yes"
-  ;;
   --enable-sunmlib) mlib="yes"
   ;;
   --enable-pthreads) pthreads="yes"
@@ -1758,10 +1740,7 @@
 fi
 echo "Sun medialib support"  $mlib
 echo "pthreads support"      $pthreads
-echo "AMR-NB float support"  $amr_nb
-echo "AMR-NB fixed support"  $amr_nb_fixed
-echo "AMR-WB float support"  $amr_wb
-echo "AMR-WB IF2 support"    $amr_if2
+echo "AMR-NB decoding support"  $amr_nb
 echo "network support      $network"
 if test "$network" = "yes" ; then
     echo "IPv6 support         $ipv6"
@@ -1942,10 +1921,6 @@
 echo "SRC_PATH_BARE=$source_path" >> config.mak
 echo "BUILD_ROOT=\"$PWD\"" >> config.mak
 
-if test "$amr_if2" = "yes" ; then
-  echo "AMR_CFLAGS=-DIF2=1" >> config.mak
-fi
-
 # Apparently it's not possible to portably echo a backslash.
 if test "$asmalign_pot" = "yes" ; then
   printf '#define ASMALIGN(ZEROBITS) ".align " #ZEROBITS "\\n\\t"\n' >> $TMPH
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 7261)
+++ libavcodec/allcodecs.c	(working copy)
@@ -169,12 +169,9 @@
 #endif
     REGISTER_ENCODER(AC3, ac3);
     REGISTER_DECODER(ALAC, alac);
-#if defined(CONFIG_AMR_NB) || defined(CONFIG_AMR_NB_FIXED)
-    REGISTER_ENCDEC (AMR_NB, amr_nb);
-#endif
-#ifdef CONFIG_AMR_WB
-    REGISTER_ENCDEC (AMR_WB, amr_wb);
-#endif
+
+    REGISTER_DECODER(AMR_NB, amr_nb);
+
     REGISTER_DECODER(COOK, cook);
     REGISTER_DECODER(DSICINAUDIO, dsicinaudio);
 #ifdef CONFIG_DTS
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 7261)
+++ libavcodec/Makefile	(working copy)
@@ -4,7 +4,7 @@
 #
 include ../config.mak
 
-CFLAGS+=-I$(SRC_PATH)/libswscale $(AMR_CFLAGS)
+CFLAGS+=-I$(SRC_PATH)/libswscale
 
 OBJS= bitstream.o \
       utils.o \
@@ -49,6 +49,7 @@
 OBJS-$(CONFIG_AASC_DECODER)            += aasc.o
 OBJS-$(CONFIG_AC3_ENCODER)             += ac3enc.o
 OBJS-$(CONFIG_ALAC_DECODER)            += alac.o
+OBJS-$(CONFIG_AMR_NB_DECODER)          += amr.o
 OBJS-$(CONFIG_ASV1_DECODER)            += asv1.o
 OBJS-$(CONFIG_ASV1_ENCODER)            += asv1.o
 OBJS-$(CONFIG_ASV2_DECODER)            += asv1.o
@@ -264,35 +265,6 @@
 # currently using libdts for dts decoding
 OBJS-$(CONFIG_DTS)                     += dtsdec.o
 
-OBJS-$(CONFIG_AMR)                     += amr.o
-OBJS-$(CONFIG_AMR_NB)                  += amr_float/sp_dec.o     \
-                                          amr_float/sp_enc.o     \
-                                          amr_float/interf_dec.o \
-                                          amr_float/interf_enc.o
-
-ifeq ($(CONFIG_AMR_NB_FIXED),yes)
-EXTRAOBJS += amr/*.o
-EXTRADEPS=amrlibs
-endif
-
-OBJS-$(CONFIG_AMR_WB)                  += amrwb_float/dec_acelp.o \
-                                          amrwb_float/dec_dtx.o   \
-                                          amrwb_float/dec_gain.o  \
-                                          amrwb_float/dec_if.o    \
-                                          amrwb_float/dec_lpc.o   \
-                                          amrwb_float/dec_main.o  \
-                                          amrwb_float/dec_rom.o   \
-                                          amrwb_float/dec_util.o  \
-                                          amrwb_float/enc_acelp.o \
-                                          amrwb_float/enc_dtx.o   \
-                                          amrwb_float/enc_gain.o  \
-                                          amrwb_float/enc_if.o    \
-                                          amrwb_float/enc_lpc.o   \
-                                          amrwb_float/enc_main.o  \
-                                          amrwb_float/enc_rom.o   \
-                                          amrwb_float/enc_util.o  \
-                                          amrwb_float/if_rom.o
-
 OBJS-$(CONFIG_AAC_PARSER)              += parser.o
 OBJS-$(CONFIG_AC3_PARSER)              += parser.o
 OBJS-$(CONFIG_CAVSVIDEO_PARSER)        += cavs.o parser.o
@@ -419,9 +391,6 @@
 
 include ../common.mak
 
-amrlibs:
-	$(MAKE) -C amr spclib fipoplib
-
 tests: apiexample $(TESTS)
 
 clean::
@@ -435,10 +404,7 @@
 	   sh4/*.o sh4/*~ \
 	   sparc/*.o sparc/*~ \
 	   liba52/*.o liba52/*~ \
-	   amr_float/*.o \
 	   apiexample $(TESTS)
-	-$(MAKE) -C amr clean
-	-$(MAKE) -C amrwb_float -f makefile.gcc clean
 
 apiexample: apiexample.o $(LIB)
 
@@ -454,4 +420,4 @@
 
 fft-test: fft-test.o $(LIB)
 
-.PHONY: amrlibs tests
+.PHONY: tests
Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(revision 7261)
+++ libavcodec/avcodec.h	(working copy)
@@ -192,7 +192,6 @@
 
     /* AMR */
     CODEC_ID_AMR_NB= 0x12000,
-    CODEC_ID_AMR_WB,
 
     /* RealAudio codecs*/
     CODEC_ID_RA_144= 0x13000,
@@ -2240,9 +2239,6 @@
 extern AVCodec vp6_decoder;
 extern AVCodec vp6f_decoder;
 extern AVCodec amr_nb_decoder;
-extern AVCodec amr_nb_encoder;
-extern AVCodec amr_wb_encoder;
-extern AVCodec amr_wb_decoder;
 extern AVCodec aac_decoder;
 extern AVCodec mpeg4aac_decoder;
 extern AVCodec asv1_decoder;
