Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 7382)
+++ libavcodec/allcodecs.c	(working copy)
@@ -175,6 +175,11 @@
 #ifdef CONFIG_AMR_WB
     REGISTER_ENCDEC (AMR_WB, amr_wb);
 #endif
+#ifndef CONFIG_AMR_NB
+#ifndef CONFIG_AMR_NB_FIXED
+    REGISTER_DECODER(FFAMR_NB, ffamr_nb);
+#endif
+#endif
     REGISTER_DECODER(COOK, cook);
     REGISTER_DECODER(DSICINAUDIO, dsicinaudio);
 #ifdef CONFIG_DTS
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 7382)
+++ libavcodec/Makefile	(working copy)
@@ -71,6 +71,7 @@
 OBJS-$(CONFIG_DVVIDEO_DECODER)         += dv.o
 OBJS-$(CONFIG_DVVIDEO_ENCODER)         += dv.o
 OBJS-$(CONFIG_EIGHTBPS_DECODER)        += 8bps.o
+OBJS-$(CONFIG_FFAMR_NB_DECODER)        += ffamr_nb.o
 OBJS-$(CONFIG_FFV1_DECODER)            += ffv1.o
 OBJS-$(CONFIG_FFV1_ENCODER)            += ffv1.o
 OBJS-$(CONFIG_FFVHUFF_DECODER)         += huffyuv.o
Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(revision 7382)
+++ libavcodec/avcodec.h	(working copy)
@@ -2241,6 +2241,7 @@
 extern AVCodec vp5_decoder;
 extern AVCodec vp6_decoder;
 extern AVCodec vp6f_decoder;
+extern AVCodec ffamr_nb_decoder;
 extern AVCodec amr_nb_decoder;
 extern AVCodec amr_nb_encoder;
 extern AVCodec amr_wb_encoder;
