Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 11162)
+++ libavcodec/Makefile	(working copy)
@@ -37,6 +36,7 @@
 OBJS-$(CONFIG_AC3_ENCODER)             += ac3enc.o ac3tab.o ac3.o
 OBJS-$(CONFIG_ALAC_DECODER)            += alac.o
 OBJS-$(CONFIG_ALAC_ENCODER)            += alacenc.o lpc.o
+OBJS-$(CONFIG_AMRNB_DECODER)           += amrnbfloatdec.o
 OBJS-$(CONFIG_AMV_DECODER)             += sp5xdec.o mjpegdec.o mjpeg.o
 OBJS-$(CONFIG_APE_DECODER)             += apedec.o
 OBJS-$(CONFIG_ASV1_DECODER)            += asv1.o mpeg12data.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 11162)
+++ libavcodec/allcodecs.c	(working copy)
@@ -177,6 +173,7 @@
     REGISTER_DECODER (AAC, aac);
     REGISTER_ENCDEC  (AC3, ac3);
     REGISTER_ENCDEC  (ALAC, alac);
+    REGISTER_DECODER (AMRNB, amrnb);
     REGISTER_DECODER (APE, ape);
     REGISTER_DECODER (ATRAC3, atrac3);
     REGISTER_DECODER (COOK, cook);
Index: doc/general.texi
===================================================================
--- doc/general.texi	(revision 16142)
+++ doc/general.texi	(working copy)
@@ -344,8 +344,8 @@
 @item AAC                    @tab  E  @tab  X
     @tab encoding supported through external library libfaac
 @item AC-3                   @tab IX  @tab IX
-@item AMR-NB                 @tab  E  @tab  E
-    @tab supported through external library libamrnb
+@item AMR-NB                 @tab  X  @tab  E
+    @tab encoding supported through external library libamrnb
 @item AMR-WB                 @tab  E  @tab  E
     @tab supported through external library libamrwb
 @item AMV IMA ADPCM          @tab     @tab  X
Index: Changelog
===================================================================
--- Changelog	(revision 16142)
+++ Changelog	(working copy)
@@ -142,6 +142,8 @@
 - Electronic Arts TGQ decoder
 - RV40 decoder
 - QCELP / PureVoice decoder
+- AMR-NB decoder
+	
 
 version 0.4.9-pre1:
 
