Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 19401)
+++ libavcodec/Makefile	(working copy)
@@ -42,6 +42,7 @@
 OBJS-$(CONFIG_AC3_ENCODER)             += ac3enc.o ac3tab.o ac3.o
 OBJS-$(CONFIG_ALAC_DECODER)            += alac.o
 OBJS-$(CONFIG_ALAC_ENCODER)            += alacenc.o lpc.o
+OBJS-$(CONFIG_AMRNB_DECODER)           += amrnbfloatdec.o qcelp_lsp.o celp_filters.o celp_math.o
 OBJS-$(CONFIG_AMV_DECODER)             += sp5xdec.o mjpegdec.o mjpeg.o
 OBJS-$(CONFIG_APE_DECODER)             += apedec.o
 OBJS-$(CONFIG_ASV1_DECODER)            += asv1.o mpeg12data.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 19401)
+++ libavcodec/allcodecs.c	(working copy)
@@ -198,6 +198,7 @@
     REGISTER_ENCDEC  (AAC, aac);
     REGISTER_ENCDEC  (AC3, ac3);
     REGISTER_ENCDEC  (ALAC, alac);
+    REGISTER_DECODER (AMRNB, amrnb);
     REGISTER_DECODER (APE, ape);
     REGISTER_DECODER (ATRAC3, atrac3);
     REGISTER_DECODER (COOK, cook);
Index: doc/general.texi
===================================================================
--- doc/general.texi	(revision 19401)
+++ doc/general.texi	(working copy)
@@ -523,8 +523,8 @@
 @item ADPCM Westwood Studios IMA @tab     @tab  X
     @tab Used in Westwood Studios games like Command and Conquer.
 @item ADPCM Yamaha           @tab  X  @tab  X
-@item AMR-NB                 @tab  E  @tab  E
-    @tab supported through external library libopencore-amrnb
+@item AMR-NB                 @tab  X  @tab  E
+    @tab encoding supported through external library libopencore-amrnb
 @item AMR-WB                 @tab     @tab  E
     @tab decoding supported through external library libopencore-amrwb
 @item Apple lossless audio   @tab  X  @tab  X
Index: Changelog
===================================================================
--- Changelog	(revision 19401)
+++ Changelog	(working copy)
@@ -28,6 +28,7 @@
 - DivX (XSUB) subtitle encoder
 - nonfree libamr support for AMR-NB/WB decoding/encoding removed
 - Experimental AAC encoder
+- AMR-NB decoder
 
 
 
Index: libavcodec/celp_filters.c
===================================================================
--- libavcodec/celp_filters.c	(revision 19401)
+++ libavcodec/celp_filters.c	(working copy)
@@ -84,6 +84,31 @@
     return 0;
 }
 
+void ff_celp_convolve_circf(
+        float* fc_out,
+        const float* fc_in,
+        const float* filter,
+        int len)
+{
+    int i, k;
+
+    memset(fc_out, 0, len * sizeof(float));
+
+    /* Since there are few pulses over an entire subframe (i.e. almost
+       all fc_in[i] are zero) it is faster to loop over fc_in first. */
+    for(i=0; i<len; i++)
+    {
+        if(fc_in[i])
+        {
+            for(k=0; k<i; k++)
+                fc_out[k] += fc_in[i] * filter[len + k - i];
+
+            for(k=i; k<len; k++)
+                fc_out[k] += fc_in[i] * filter[      k - i];
+        }
+    }
+}
+
 void ff_celp_lp_synthesis_filterf(
         float *out,
         const float* filter_coeffs,
@@ -113,13 +138,10 @@
 {
     int i,n;
 
-    // Avoids a +1 in the inner loop.
-    filter_length++;
-
     for(n=0; n<buffer_length; n++)
     {
-        out[n] = in[n];
-        for(i=1; i<filter_length; i++)
-            out[n] -= filter_coeffs[i-1] * in[n-i];
+        out[n] = 0;
+        for(i=0; i<filter_length; i++)
+            out[n] += filter_coeffs[i] * in[n-i];
     }
 }
Index: libavcodec/celp_filters.h
===================================================================
--- libavcodec/celp_filters.h	(revision 19401)
+++ libavcodec/celp_filters.h	(working copy)
@@ -70,14 +70,31 @@
         int rounder);
 
 /**
+ * Circularly convolve fixed vector with a phase dispersion impulse
+ *        response filter (D.6.2 of G.729 and 6.1.5 of AMR).
+ * @param fc_out vector with filter applied
+ * @param fc_in source vector
+ * @param filter phase filter coefficients
+ *
+ *  fc_out[n] = sum(i,0,len-1){ fc_in[i] * filter[(len + n - i)%len] }
+ *
+ * \note fc_in and fc_out should not overlap!
+ */
+void ff_celp_convolve_circf(
+        float* fc_out,
+        const float* fc_in,
+        const float* filter,
+        int len);
+
+/**
  * LP synthesis filter.
  * @param out [out] pointer to output buffer
  *        - the array out[-filter_length, -1] must
  *        contain the previous result of this filter
- * @param filter_coeffs filter coefficients.
+ * @param filter_coeffs filter coefficients not including z^0
  * @param in input signal
  * @param buffer_length amount of data to process
- * @param filter_length filter length (10 for 10th order LP filter)
+ * @param filter_length filter_coeffs length (10 for 10th order LP filter)
  *
  * @note Output buffer must contain filter_length samples of past
  *       speech data before pointer.
@@ -94,16 +111,13 @@
 /**
  * LP zero synthesis filter.
  * @param out [out] pointer to output buffer
- * @param filter_coeffs filter coefficients.
+ * @param filter_coeffs filter coefficients including z^0 coefficient
  * @param in input signal
- *        - the array in[-filter_length, -1] must
+ *        - the array in[-filter_length+1, -1] must
  *        contain the previous input of this filter
  * @param buffer_length amount of data to process
- * @param filter_length filter length (10 for 10th order LP filter)
+ * @param filter_length filter_coeffs length (11 for 10th order LP filter)
  *
- * @note Output buffer must contain filter_length samples of past
- *       speech data before pointer.
- *
  * Routine applies A(z) filter to given speech data.
  */
 void ff_celp_lp_zero_synthesis_filterf(
