Index: libavformat/Makefile
===================================================================
--- libavformat/Makefile	(wersja 19606)
+++ libavformat/Makefile	(kopia robocza)
@@ -199,6 +199,7 @@
 OBJS-$(CONFIG_SOL_DEMUXER)               += sol.o raw.o
 OBJS-$(CONFIG_SOX_DEMUXER)               += soxdec.o raw.o
 OBJS-$(CONFIG_SOX_MUXER)                 += soxenc.o
+OBJS-$(CONFIG_SPDIF_MUXER)               += spdif.o
 OBJS-$(CONFIG_STR_DEMUXER)               += psxstr.o
 OBJS-$(CONFIG_SWF_DEMUXER)               += swfdec.o
 OBJS-$(CONFIG_SWF_MUXER)                 += swfenc.o
Index: libavformat/allformats.c
===================================================================
--- libavformat/allformats.c	(wersja 19606)
+++ libavformat/allformats.c	(kopia robocza)
@@ -178,6 +178,7 @@
     REGISTER_DEMUXER  (SMACKER, smacker);
     REGISTER_DEMUXER  (SOL, sol);
     REGISTER_MUXDEMUX (SOX, sox);
+    REGISTER_MUXER    (SPDIF, spdif);
     REGISTER_DEMUXER  (STR, str);
     REGISTER_MUXDEMUX (SWF, swf);
     REGISTER_MUXER    (TG2, tg2);
