Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 9172)
+++ libavcodec/Makefile	(working copy)
@@ -35,6 +35,7 @@
 
 HEADERS = avcodec.h opt.h
 
+OBJS-$(CONFIG_AAC_DECODER)             += aac.o
 OBJS-$(CONFIG_AASC_DECODER)            += aasc.o
 OBJS-$(CONFIG_AC3_ENCODER)             += ac3enc.o ac3tab.o ac3.o
 OBJS-$(CONFIG_ALAC_DECODER)            += alac.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 9172)
+++ libavcodec/allcodecs.c	(working copy)
@@ -164,6 +164,7 @@
     REGISTER_ENCDEC (ZMBV, zmbv);
 
     /* audio codecs */
+    REGISTER_DECODER(AAC, aac);
     REGISTER_DECODER(MPEG4AAC, mpeg4aac);
     REGISTER_ENCODER(AC3, ac3);
     REGISTER_DECODER(ALAC, alac);
Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(revision 9172)
+++ libavcodec/avcodec.h	(working copy)
@@ -2240,6 +2240,7 @@
 extern AVCodec wmv2_encoder;
 extern AVCodec zmbv_encoder;
 
+extern AVCodec aac_decoder;
 extern AVCodec aasc_decoder;
 extern AVCodec alac_decoder;
 extern AVCodec asv1_decoder;
