Index: libavcodec/ac3_parser.c
===================================================================
--- libavcodec/ac3_parser.c	(wersja 10118)
+++ libavcodec/ac3_parser.c	(kopia robocza)
@@ -84,7 +84,7 @@
     return 0;
 }
 
-static int ac3_sync(const uint8_t *buf, int *channels, int *sample_rate,
+static int ac3_sync(AVCodecContext *avctx, const uint8_t *buf, int *channels, int *sample_rate,
                     int *bit_rate, int *samples)
 {
     int err;
@@ -100,12 +100,16 @@
 
     bsid = hdr.bsid;
     if(bsid <= 10) {             /* Normal AC-3 */
+        if(avctx->codec_id == CODEC_ID_EAC3)
+            avctx->codec_id = CODEC_ID_AC3;
         *sample_rate = hdr.sample_rate;
         *bit_rate = hdr.bit_rate;
         *channels = hdr.channels;
         *samples = AC3_FRAME_SIZE;
         return hdr.frame_size;
     } else if (bsid > 10 && bsid <= 16) { /* Enhanced AC-3 */
+        if(avctx->codec_id == CODEC_ID_AC3)
+            avctx->codec_id = CODEC_ID_EAC3;
         init_get_bits(&bits, &buf[2], (AC3_HEADER_SIZE-2) * 8);
         strmtyp = get_bits(&bits, 2);
         substreamid = get_bits(&bits, 3);
@@ -152,6 +156,13 @@
     return 0;
 }
 
+AVCodecParser eac3_parser = {
+    { CODEC_ID_EAC3 },
+    sizeof(AACAC3ParseContext),
+    ac3_parse_init,
+    ff_aac_ac3_parse,
+    NULL,
+};
 
 AVCodecParser ac3_parser = {
     { CODEC_ID_AC3 },
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(wersja 10118)
+++ libavcodec/Makefile	(kopia robocza)
@@ -63,6 +63,7 @@
 OBJS-$(CONFIG_DVVIDEO_DECODER)         += dv.o
 OBJS-$(CONFIG_DVVIDEO_ENCODER)         += dv.o
 OBJS-$(CONFIG_DXA_DECODER)             += dxa.o
+OBJS-$(CONFIG_EAC3_DECODER)            += eac3dec.o ac3dec.o
 OBJS-$(CONFIG_EIGHTBPS_DECODER)        += 8bps.o
 OBJS-$(CONFIG_FFV1_DECODER)            += ffv1.o rangecoder.o golomb.o
 OBJS-$(CONFIG_FFV1_ENCODER)            += ffv1.o rangecoder.o
Index: libavcodec/aac_ac3_parser.c
===================================================================
--- libavcodec/aac_ac3_parser.c	(wersja 10118)
+++ libavcodec/aac_ac3_parser.c	(kopia robocza)
@@ -48,7 +48,7 @@
             s->inbuf_ptr += len;
             buf_size -= len;
             if ((s->inbuf_ptr - s->inbuf) == s->header_size) {
-                len = s->sync(s->inbuf, &channels, &sample_rate, &bit_rate,
+                len = s->sync(avctx, s->inbuf, &channels, &sample_rate, &bit_rate,
                               &samples);
                 if (len == 0) {
                     /* no sync found : move by one byte (inefficient, but simple!) */
Index: libavcodec/aac_ac3_parser.h
===================================================================
--- libavcodec/aac_ac3_parser.h	(wersja 10118)
+++ libavcodec/aac_ac3_parser.h	(kopia robocza)
@@ -30,7 +30,7 @@
     uint8_t *inbuf_ptr;
     int frame_size;
     int header_size;
-    int (*sync)(const uint8_t *buf, int *channels, int *sample_rate,
+    int (*sync)(AVCodecContext *avctx, const uint8_t *buf, int *channels, int *sample_rate,
                 int *bit_rate, int *samples);
     uint8_t inbuf[8192]; /* input buffer */
 } AACAC3ParseContext;
Index: libavcodec/aac_parser.c
===================================================================
--- libavcodec/aac_parser.c	(wersja 10118)
+++ libavcodec/aac_parser.c	(kopia robocza)
@@ -38,7 +38,7 @@
 };
 
 
-static int aac_sync(const uint8_t *buf, int *channels, int *sample_rate,
+static int aac_sync(AVCodecContext *avctx, const uint8_t *buf, int *channels, int *sample_rate,
                     int *bit_rate, int *samples)
 {
     GetBitContext bits;
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(wersja 10118)
+++ libavcodec/allcodecs.c	(kopia robocza)
@@ -172,6 +172,7 @@
     REGISTER_DECODER(COOK, cook);
     REGISTER_DECODER(DCA, dca);
     REGISTER_DECODER(DSICINAUDIO, dsicinaudio);
+    REGISTER_DECODER(EAC3, eac3);
     REGISTER_ENCDEC (FLAC, flac);
     REGISTER_DECODER(IMC, imc);
     REGISTER_ENCDEC (LIBAMR_NB, libamr_nb);
@@ -266,6 +267,7 @@
     REGISTER_PARSER (DCA, dca);
     REGISTER_PARSER (DVBSUB, dvbsub);
     REGISTER_PARSER (DVDSUB, dvdsub);
+    REGISTER_PARSER (EAC3, eac3);
     REGISTER_PARSER (H261, h261);
     REGISTER_PARSER (H263, h263);
     REGISTER_PARSER (H264, h264);
Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(wersja 10118)
+++ libavcodec/avcodec.h	(kopia robocza)
@@ -260,6 +260,7 @@
     CODEC_ID_GSM_MS, /* as found in WAV */
     CODEC_ID_ATRAC3,
     CODEC_ID_VOXWARE,
+    CODEC_ID_EAC3,
 
     /* subtitle codecs */
     CODEC_ID_DVD_SUBTITLE= 0x17000,
@@ -2780,6 +2781,7 @@
 extern AVCodecParser dca_parser;
 extern AVCodecParser dvbsub_parser;
 extern AVCodecParser dvdsub_parser;
+extern AVCodecParser eac3_parser;
 extern AVCodecParser h261_parser;
 extern AVCodecParser h263_parser;
 extern AVCodecParser h264_parser;
Index: libavformat/raw.c
===================================================================
--- libavformat/raw.c	(wersja 10118)
+++ libavformat/raw.c	(kopia robocza)
@@ -489,6 +489,19 @@
     .extensions = "ac3",
 };
 #endif
+#ifdef CONFIG_EAC3_DEMUXER
+AVInputFormat eac3_demuxer = {
+    "eac3",
+    "raw ac3",
+    0,
+    ac3_probe,
+    ac3_read_header,
+    raw_read_partial_packet,
+    raw_read_close,
+    .flags= AVFMT_GENERIC_INDEX,
+    .extensions = "ac3",
+};
+#endif
 
 #ifdef CONFIG_MUXERS
 AVOutputFormat ac3_muxer = {
Index: libavcodec/allcodecs.h
===================================================================
--- libavcodec/allcodecs.h	(wersja 10118)
+++ libavcodec/allcodecs.h	(kopia robocza)
@@ -98,6 +98,7 @@
 extern AVCodec dsicinvideo_decoder;
 extern AVCodec dvvideo_decoder;
 extern AVCodec dxa_decoder;
+extern AVCodec eac3_decoder;
 extern AVCodec eightbps_decoder;
 extern AVCodec ffv1_decoder;
 extern AVCodec ffvhuff_decoder;
Index: libavcodec/ac3enc.c
===================================================================
--- libavcodec/ac3enc.c	(wersja 10118)
+++ libavcodec/ac3enc.c	(kopia robocza)
@@ -479,7 +479,8 @@
         for(ch=0;ch<s->nb_all_channels;ch++) {
             ff_ac3_bit_alloc_calc_bap(mask[i][ch], psd[i][ch], 0,
                                       s->nb_coefs[ch], snroffset,
-                                      s->bit_alloc.floor, bap[i][ch]);
+                                      s->bit_alloc.floor, ff_ac3_baptab,
+                                      bap[i][ch]);
             frame_bits += compute_mantissa_size(s, bap[i][ch],
                                                  s->nb_coefs[ch]);
         }
