Index: libavcodec/ac3dec.h
===================================================================
--- libavcodec/ac3dec.h	(revision 14773)
+++ libavcodec/ac3dec.h	(working copy)
@@ -169,4 +169,16 @@
 ///@}
 } AC3DecodeContext;
 
+/**
+ * Parse the E-AC-3 frame header.
+ * This parses both the bit stream info and audio frame header.
+ */
+int ff_eac3_parse_header(AC3DecodeContext *s);
+
+/**
+ * Decode mantissas in a single channel for the entire frame.
+ * This is used when AHT mode is enabled.
+ */
+void ff_eac3_get_transform_coeffs_aht_ch(AC3DecodeContext *s, int ch);
+
 #endif /* FFMPEG_AC3DEC_H */
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 14773)
+++ libavcodec/Makefile	(working copy)
@@ -26,7 +26,7 @@
 OBJS-$(CONFIG_ENCODERS)                += faandct.o jfdctfst.o jfdctint.o
 
 OBJS-$(CONFIG_AASC_DECODER)            += aasc.o
-OBJS-$(CONFIG_AC3_DECODER)             += ac3dec.o ac3tab.o ac3dec_data.o ac3.o mdct.o fft.o
+OBJS-$(CONFIG_AC3_DECODER)             += eac3dec.o ac3dec.o ac3tab.o ac3dec_data.o ac3.o mdct.o fft.o
 OBJS-$(CONFIG_AC3_ENCODER)             += ac3enc.o ac3tab.o ac3.o
 OBJS-$(CONFIG_ALAC_DECODER)            += alac.o
 OBJS-$(CONFIG_AMV_DECODER)             += sp5xdec.o mjpegdec.o mjpeg.o
Index: libavcodec/ac3enc.c
===================================================================
--- libavcodec/ac3enc.c	(revision 14773)
+++ libavcodec/ac3enc.c	(working copy)
@@ -1365,5 +1365,5 @@
     AC3_encode_close,
     NULL,
     .sample_fmts = (enum SampleFormat[]){SAMPLE_FMT_S16,SAMPLE_FMT_NONE},
-    .long_name = NULL_IF_CONFIG_SMALL("ATSC A/52 / AC-3"),
+    .long_name = NULL_IF_CONFIG_SMALL("ATSC A/52 (AC-3, E-AC-3)"),
 };
