Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile
+++ libavcodec/Makefile
@@ -151,6 +151,8 @@
 OBJS-$(CONFIG_RV10_ENCODER)            += rv10.o mpegvideo_enc.o motion_est.o ratecontrol.o h263.o
 OBJS-$(CONFIG_RV20_DECODER)            += rv10.o h263.o
 OBJS-$(CONFIG_RV20_ENCODER)            += rv10.o mpegvideo_enc.o motion_est.o ratecontrol.o h263.o
+OBJS-$(CONFIG_RV30_DECODER)            += rv30.o rv34.o h264pred.o
+OBJS-$(CONFIG_RV40_DECODER)            += rv40.o rv34.o h264pred.o
 OBJS-$(CONFIG_SGI_DECODER)             += sgidec.o
 OBJS-$(CONFIG_SGI_ENCODER)             += sgienc.o rle.o
 OBJS-$(CONFIG_SHORTEN_DECODER)         += shorten.o golomb.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c
+++ libavcodec/allcodecs.c
@@ -128,6 +128,8 @@
     REGISTER_DECODER(RPZA, rpza);
     REGISTER_ENCDEC (RV10, rv10);
     REGISTER_ENCDEC (RV20, rv20);
+    REGISTER_DECODER(RV30, rv30);
+    REGISTER_DECODER(RV40, rv40);
     REGISTER_ENCDEC (SGI, sgi);
     REGISTER_DECODER(SMACKER, smacker);
     REGISTER_DECODER(SMC, smc);
Index: libavformat/riff.c
===================================================================
--- libavformat/riff.c
+++ libavformat/riff.c
@@ -172,6 +172,8 @@
     { CODEC_ID_JPEG2000, MKTAG('M', 'J', '2', 'C') },
     { CODEC_ID_VMNC, MKTAG('V', 'M', 'n', 'c') },
     { CODEC_ID_TARGA, MKTAG('t', 'g', 'a', ' ') },
+    { CODEC_ID_RV30, MKTAG('R', 'V', '3', '0') },
+    { CODEC_ID_RV40, MKTAG('R', 'V', '4', '0') },
     { CODEC_ID_NONE, 0 },
 };
 
