diff --git a/libavcodec/utils.c b/libavcodec/utils.c
index 43147a5..c9ec405 100644
--- a/libavcodec/utils.c
+++ b/libavcodec/utils.c
@@ -672,7 +672,7 @@ int avcodec_close(AVCodecContext *avctx)
 
     if (HAVE_THREADS && avctx->thread_opaque)
         avcodec_thread_free(avctx);
-    if (avctx->codec->close)
+    if (avctx->codec && avctx->codec->close)
         avctx->codec->close(avctx);
     avcodec_default_free_buffers(avctx);
     av_freep(&avctx->priv_data);
