From a990d5415ca2933a936cab3f0907bf47b6d4bf4e Mon Sep 17 00:00:00 2001
From: David Conrad <davedc@Crescens.local>
Date: Wed, 25 Jul 2007 16:33:04 -0400
Subject: [PATCH] Add segment UID EBML ID

---
 libavformat/matroska.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/libavformat/matroska.h b/libavformat/matroska.h
index 98b2165..a683bc7 100644
--- a/libavformat/matroska.h
+++ b/libavformat/matroska.h
@@ -64,6 +64,7 @@
 #define MATROSKA_ID_WRITINGAPP 0x5741
 #define MATROSKA_ID_MUXINGAPP  0x4D80
 #define MATROSKA_ID_DATEUTC    0x4461
+#define MATROSKA_ID_SEGMENTUID 0x73A4
 
 /* ID in the tracks master */
 #define MATROSKA_ID_TRACKENTRY 0xAE
-- 
1.5.1.4

