diff --git a/libavformat/Makefile b/libavformat/Makefile
index 7c837ce..dacbe68 100644
--- a/libavformat/Makefile
+++ b/libavformat/Makefile
@@ -72,6 +72,7 @@ OBJS-$(CONFIG_IPMOVIE_DEMUXER)           += ipmovie.o
 OBJS-$(CONFIG_M4V_DEMUXER)               += raw.o
 OBJS-$(CONFIG_M4V_MUXER)                 += raw.o
 OBJS-$(CONFIG_MATROSKA_DEMUXER)          += matroskadec.o matroska.o riff.o
+OBJS-$(CONFIG_MATROSKA_MUXER)            += matroskaenc.o matroska.o riff.o
 OBJS-$(CONFIG_MJPEG_DEMUXER)             += raw.o
 OBJS-$(CONFIG_MJPEG_MUXER)               += raw.o
 OBJS-$(CONFIG_MM_DEMUXER)                += mm.o
diff --git a/libavformat/allformats.c b/libavformat/allformats.c
index 84ba067..0b4b4af 100644
--- a/libavformat/allformats.c
+++ b/libavformat/allformats.c
@@ -92,7 +92,7 @@ void av_register_all(void)
     if (!ENABLE_NUT_DEMUXER) REGISTER_DEMUXER (LIBNUT, libnut);
     REGISTER_MUXER   (LIBNUT, libnut);
     REGISTER_MUXDEMUX(M4V, m4v);
-    REGISTER_DEMUXER (MATROSKA, matroska);
+    REGISTER_MUXDEMUX(MATROSKA, matroska);
     REGISTER_MUXDEMUX(MJPEG, mjpeg);
     REGISTER_DEMUXER (MM, mm);
     REGISTER_MUXDEMUX(MMF, mmf);
diff --git a/libavformat/allformats.h b/libavformat/allformats.h
index 4c9a904..c1d4e6a 100644
--- a/libavformat/allformats.h
+++ b/libavformat/allformats.h
@@ -141,6 +141,7 @@ extern AVOutputFormat image_muxer;
 extern AVOutputFormat imagepipe_muxer;
 extern AVOutputFormat libnut_muxer;
 extern AVOutputFormat m4v_muxer;
+extern AVOutputFormat matroska_muxer;
 extern AVOutputFormat mjpeg_muxer;
 extern AVOutputFormat mmf_muxer;
 extern AVOutputFormat mov_muxer;
