Index: libavformat/asf.c
===================================================================
--- libavformat/asf.c	(revision 24076)
+++ libavformat/asf.c	(working copy)
@@ -110,6 +110,10 @@
         0x01, 0xCD, 0x87, 0xF4, 0x51, 0xA9, 0xCF, 0x11, 0x8E, 0xE6, 0x00, 0xC0, 0x0C, 0x20, 0x53, 0x65
 };
 
+const ff_asf_guid ff_asf_stream_bitrate_properties = {
+        0xce, 0x75, 0xf8, 0x7b, 0x8d, 0x46, 0xd1, 0x11, 0x8d, 0x82, 0x00, 0x60, 0x97, 0xc9, 0xa2, 0xb2
+};
+
 /* I am not a number !!! This GUID is the one found on the PC used to
    generate the stream */
 const ff_asf_guid ff_asf_my_guid = {
Index: libavformat/asf.h
===================================================================
--- libavformat/asf.h	(revision 24076)
+++ libavformat/asf.h	(working copy)
@@ -164,6 +164,7 @@
 extern const ff_asf_guid ff_asf_content_encryption;
 extern const ff_asf_guid ff_asf_ext_content_encryption;
 extern const ff_asf_guid ff_asf_digital_signature;
+extern const ff_asf_guid ff_asf_stream_bitrate_properties;
 
 extern const AVMetadataConv ff_asf_metadata_conv[];
 
