Index: riff.c
===================================================================
--- riff.c	(wersja 9451)
+++ riff.c	(kopia robocza)
@@ -169,6 +169,7 @@
     { CODEC_ID_ZMBV, MKTAG('Z', 'M', 'B', 'V') },
     { CODEC_ID_KMVC, MKTAG('K', 'M', 'V', 'C') },
     { CODEC_ID_CAVS, MKTAG('C', 'A', 'V', 'S') },
+    { CODEC_ID_JPEG2000, MKTAG('m', 'j', 'p', '2') },
     { CODEC_ID_JPEG2000, MKTAG('M', 'J', '2', 'C') },
     { CODEC_ID_VMNC, MKTAG('V', 'M', 'n', 'c') },
     { CODEC_ID_TARGA, MKTAG('t', 'g', 'a', ' ') },
