Index: Changelog
===================================================================
--- Changelog	(revision 14483)
+++ Changelog	(working copy)
@@ -127,6 +127,7 @@
 - Motion Pixels Video decoder
 - Motion Pixels MVI demuxer
 - removed animated GIF decoder/demuxer
+- JPEG2000 encoder and decoder
 
 version 0.4.9-pre1:
 
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 14483)
+++ libavcodec/Makefile	(working copy)
@@ -99,6 +99,8 @@
 OBJS-$(CONFIG_INTERPLAY_VIDEO_DECODER) += interplayvideo.o
 OBJS-$(CONFIG_JPEGLS_DECODER)          += jpeglsdec.o jpegls.o mjpegdec.o mjpeg.o golomb.o
 OBJS-$(CONFIG_JPEGLS_ENCODER)          += jpeglsenc.o jpegls.o golomb.o
+OBJS-$(CONFIG_JPEG2000_DECODER)        += j2kdec.o mqcdec.o mqc.o j2k.o dwt.o
+OBJS-$(CONFIG_JPEG2000_ENCODER)        += j2kenc.o mqcenc.o mqc.o j2k.o dwt.o
 OBJS-$(CONFIG_KMVC_DECODER)            += kmvc.o
 OBJS-$(CONFIG_LJPEG_ENCODER)           += ljpegenc.o mjpegenc.o mjpeg.o mpegvideo_enc.o motion_est.o ratecontrol.o mpeg12data.o mpegvideo.o
 OBJS-$(CONFIG_LOCO_DECODER)            += loco.o golomb.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 14483)
+++ libavcodec/allcodecs.c	(working copy)
@@ -101,6 +101,7 @@
     REGISTER_DECODER (INDEO2, indeo2);
     REGISTER_DECODER (INDEO3, indeo3);
     REGISTER_DECODER (INTERPLAY_VIDEO, interplay_video);
+    REGISTER_ENCDEC  (JPEG2000, jpeg2000);
     REGISTER_ENCDEC  (JPEGLS, jpegls);
     REGISTER_DECODER (KMVC, kmvc);
     REGISTER_ENCODER (LJPEG, ljpeg);
Index: doc/general.texi
===================================================================
--- doc/general.texi	(revision 14483)
+++ doc/general.texi	(working copy)
@@ -253,6 +253,7 @@
     @tab Used in Interplay .MVE files.
 @item JPEG-LS                @tab  X  @tab  X
     @tab fourcc: MJLS, lossless and near-lossless is supported.
+@item JPEG2000               @tab  X  @tab  X
 @item KMVC                   @tab     @tab  X
     @tab Codec used in Worms games.
 @item LOCO                   @tab     @tab  X
Index: libavformat/riff.c
===================================================================
--- libavformat/riff.c	(revision 14483)
+++ libavformat/riff.c	(working copy)
@@ -156,6 +156,7 @@
     { CODEC_ID_ZMBV,         MKTAG('Z', 'M', 'B', 'V') },
     { CODEC_ID_KMVC,         MKTAG('K', 'M', 'V', 'C') },
     { CODEC_ID_CAVS,         MKTAG('C', 'A', 'V', 'S') },
+    { CODEC_ID_JPEG2000,     MKTAG('m', 'j', 'p', '2') },
     { CODEC_ID_JPEG2000,     MKTAG('M', 'J', '2', 'C') },
     { CODEC_ID_VMNC,         MKTAG('V', 'M', 'n', 'c') },
     { CODEC_ID_TARGA,        MKTAG('t', 'g', 'a', ' ') },
