From a507defd145d6a0c0c4ff96770795461db6c0b7b Mon Sep 17 00:00:00 2001
From: Geza Kovacs <gkovacs@mit.edu>
Date: Thu, 2 Apr 2009 04:41:17 -0400
Subject: [PATCH] removed color table check

---
 ffmpeg.c |   10 ----------
 1 files changed, 0 insertions(+), 10 deletions(-)

diff --git a/ffmpeg.c b/ffmpeg.c
index cda0d50..6beb890 100644
--- a/ffmpeg.c
+++ b/ffmpeg.c
@@ -1546,8 +1546,6 @@ static int stream_index_from_inputs(AVFormatContext **input_files,
  */
 static int check_same_settings(AVFormatContext **input_format_contexts, int num_input_files)
 {
-    int color_table_id;
-    char color_table_id_set = 0;
     int width;
     char width_set = 0;
     int height;
@@ -1563,14 +1561,6 @@ static int check_same_settings(AVFormatContext **input_format_contexts, int num_
                 fprintf(stderr, "Error: could not open codec for input file %s\n", input_format_contexts[i]->filename);
                 return -1;
             }
-            if (!color_table_id_set) {
-                color_table_id_set = 1;
-                color_table_id = input_format_contexts[i]->streams[j]->codec->color_table_id;
-            }
-            else if (input_format_contexts[i]->streams[j]->codec->color_table_id != color_table_id) {
-                fprintf(stderr, "Error: different color table for input file %s: %i vs existing %i \n", input_format_contexts[i]->filename, input_format_contexts[i]->streams[j]->codec->color_table_id, color_table_id);
-                return -1;
-            }
             if (!frame_rate_set) {
                 frame_rate_set = 1;
                 frame_rate_num = input_format_contexts[i]->streams[j]->r_frame_rate.num;
-- 
1.6.0.4

