diff --git a/ffmpeg.c b/ffmpeg.c
index e899180..e5f27cd 100644
--- a/ffmpeg.c
+++ b/ffmpeg.c
@@ -1237,7 +1237,8 @@ static void print_report(AVFormatContext **output_files,
 /* pkt = NULL means EOF (needed to flush decoder buffers) */
 static int output_packet(AVInputStream *ist, int ist_index,
                          AVOutputStream **ost_table, int nb_ostreams,
-                         const AVPacket *pkt)
+                         const AVPacket *pkt,
+                         AVFormatContext *is)
 {
     AVFormatContext *os;
     AVOutputStream *ost;
@@ -1250,7 +1251,24 @@ static int output_packet(AVInputStream *ist, int ist_index,
     AVSubtitle subtitle, *subtitle_to_free;
     int got_subtitle;
     AVPacket avpkt;
-
+    if (ist && is && pkt && is->iformat && is->iformat->long_name &&
+        !strncmp(is->iformat->long_name, "CONCAT", 6) && is->nb_streams > pkt->stream_index &&
+        is->streams && is->streams[pkt->stream_index] && is->streams[pkt->stream_index]->codec) {
+        ist->st = is->streams[pkt->stream_index];
+        if (!ist->st->codec->codec) {
+            AVCodec *codec = avcodec_find_decoder(ist->st->codec->codec_id);
+            if (!codec) {
+                av_log(ist->st->codec, AV_LOG_ERROR, "Decoder (codec id %d) not found for input stream #%d.%d\n",
+                       ist->st->codec->codec_id, ist->file_index, ist->index);
+                return AVERROR(EINVAL);
+             }
+             if (avcodec_open(ist->st->codec, codec) < 0) {
+                av_log(ist->st->codec, AV_LOG_ERROR, "Error while opening decoder for input stream #%d.%d\n",
+                       ist->file_index, ist->index);
+                return AVERROR(EINVAL);
+             }
+         }
+    }
     if(ist->next_pts == AV_NOPTS_VALUE)
         ist->next_pts= ist->pts;
 
@@ -2278,8 +2296,7 @@ static int av_encode(AVFormatContext **output_files,
         }
 
         //fprintf(stderr,"read #%d.%d size=%d\n", ist->file_index, ist->index, pkt.size);
-        if (output_packet(ist, ist_index, ost_table, nb_ostreams, &pkt) < 0) {
-
+        if (output_packet(ist, ist_index, ost_table, nb_ostreams, &pkt, is) < 0) {
             if (verbose >= 0)
                 fprintf(stderr, "Error while decoding stream #%d.%d\n",
                         ist->file_index, ist->index);
@@ -2300,7 +2317,7 @@ static int av_encode(AVFormatContext **output_files,
     for(i=0;i<nb_istreams;i++) {
         ist = ist_table[i];
         if (ist->decoding_needed) {
-            output_packet(ist, i, ost_table, nb_ostreams, NULL);
+            output_packet(ist, i, ost_table, nb_ostreams, NULL, is);
         }
     }
 
@@ -2892,6 +2909,7 @@ static void opt_input_file(const char *filename)
         print_error(filename, err);
         av_exit(1);
     }
+
     if(opt_programid) {
         int i;
         for(i=0; i<ic->nb_programs; i++)
